Seecer Oases Sailfish (SOS) version 0.1
----------------------------
Dilp A Durai (ddurai@contact.mmci.uni-saarland.de)
Marcel H Schulz (mschulz@mmci.uni-saarland.de)


Requirements
----------------------------
for SOS
	1)	64-bit linux operating system. 
	2)	A physical memory of around 12 GB is recommended.
	3)	Jellyfish version 2.0
for SEECER
 	1)	GNU Scientific Library
for SAILFISH
	1)	Any version of [g++](gcc.gnu.org) >= 4.7


Downloading
-----------------------------
The pipeline can be downloaded using the following command:
	git clone https://github.com/SchulzLab/SOS
The downloaded folder should have the following files:
	pipeline.py
	sequence_cleaner.py
	install_script.py	
	export.sh
	config.txt	
	Jellyfish-2.0.0.tar
	Sample_Data (folder)

Installation
-----------------------------
1)	After downloading the SOS distribution, change into the directory
		cd SOS

2)	Then run the python script install_script.py to install the softwares (SEECER,OASES and SAILFISH)
		python install_script.py -f <destination_folder>
	This should install SEECER, OASES and SAILFISH in the provided destination folder. 
	Note: Running the script is optional. The users can download and install the softwares seperately  

3)	The PATH variable should contain executables of each software and the LD_LIBRARY_PATH variable should contain 'lib' folder of Sailfish.
	To make this process easier we have included a bash script export.sh. To use this script type the following command:
		source export.sh <folder_name>
	Note: To make use of this script all the softwares should be installed in the same directory (folder_name)

4)	To install jellyfish-2.0.0 unpack the .tar file
		tar -xvf jellyfish-2.0.0.tar
	change to the directory 
		cd jellyfish-2.0.0
	complie jellyfish
		./configure
		make
	set the PATH variable 
		export PATH=/install dir/:$PATH  

Usage
----------------------------
To see the options available with the pipeline run the following command:
	python pipeline.py --help

	-h, --help            show this help message and exit
  	-o FOLDERNAME, --output=FOLDERNAME
        	                Output directory
  	-i INPUTNAME, --input=INPUTNAME
        	                Single end reads (comma seperated)
  	-l LEFT_NAME, --left=LEFT_NAME
        	                Left end reads (comma seperated)
  	-r RIGHT_NAME, --right=RIGHT_NAME
        	                Right end reads (comma seperated)
  	-t THRESHOLD, --threshold=THRESHOLD
        	                Threshold value for file reduction (optional)
  	-p PERC, --per=PERC   Percentage value for file reduction(optional)
  	-c CONFIG_FILE, --config=CONFIG_FILE
        	                Config file (required)
  	-s STEP, --step=STEP  The step you want start with (1-SEECER(default)
        	                2-OASES 3-SAILFISH)

Sample command : 
	for single end : python pipeline.py -i Sample_Data/SRR027877-small-p1.fastq -c config_s.txt -o Sample_Output
	for paired end : python pipeline.py -l Sample_Data/SRR027877-small-p1.fastq -r Sample_Data/SRR027877-small-p2.fastq -c config_p.txt -o 		Sample_Output

Output Folder
The output folder will have a folders generated by the oases run and the folders generated by the sailfish run. 
It will also have a folder named Final_Output which will contain the following file :
	1)	reads_corrected.fa (result of SEECER)
	2)	transcripts.fa (result of OASES)
	3)	quant.sf(result of SAILFISH)
If the read reduction option is used then the output folder will contain folder named Jellyfish_Output which stores the intermediate results generated by the read reduction step.


Enhancements in future versions
-----------------------------
The pipeline will include an algorithm for predicting the kmer range for error correction and assembly step.      
